File: corebank-backend/corebank/models/investment.py
"""
Investment-related Pydantic models for CoreBank API.

This module defines all the request/response models for investment
and wealth management functionality.
"""

from datetime import datetime, date
from decimal import Decimal
from enum import Enum
from typing import Optional, Dict, Any, List
from uuid import UUID

from pydantic import BaseModel, Field, validator


class ProductType(str, Enum):
    """Investment product types."""
    MONEY_FUND = "money_fund"
    FIXED_TERM = "fixed_term"
    MUTUAL_FUND = "mutual_fund"
    INSURANCE = "insurance"


class RiskLevel(int, Enum):
    """Risk levels for investment products."""
    VERY_LOW = 1
    LOW = 2
    MEDIUM = 3
    HIGH = 4
    VERY_HIGH = 5


class InvestmentExperience(str, Enum):
    """Investment experience levels."""
    BEGINNER = "beginner"
    INTERMEDIATE = "intermediate"
    ADVANCED = "advanced"


class InvestmentGoal(str, Enum):
    """Investment goals."""
    WEALTH_PRESERVATION = "wealth_preservation"
    STEADY_GROWTH = "steady_growth"
    AGGRESSIVE_GROWTH = "aggressive_growth"


class InvestmentHorizon(str, Enum):
    """Investment time horizons."""
    SHORT_TERM = "short_term"    # < 1 year
    MEDIUM_TERM = "medium_term"  # 1-5 years
    LONG_TERM = "long_term"      # > 5 years


class TransactionType(str, Enum):
    """Investment transaction types."""
    PURCHASE = "purchase"
    REDEMPTION = "redemption"
    DIVIDEND = "dividend"
    INTEREST = "interest"


class TransactionStatus(str, Enum):
    """Investment transaction status."""
    PENDING = "pending"
    CONFIRMED = "confirmed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class HoldingStatus(str, Enum):
    """Investment holding status."""
    ACTIVE = "active"
    MATURED = "matured"
    REDEEMED = "redeemed"


# Base models
class InvestmentProductBase(BaseModel):
    """Base model for investment products."""
    
    product_code: str = Field(..., max_length=50, description="Product code")
    name: str = Field(..., max_length=200, description="Product name")
    product_type: ProductType = Field(..., description="Product type")
    risk_level: RiskLevel = Field(..., description="Risk level (1-5)")
    expected_return_rate: Optional[Decimal] = Field(
        None, 
        max_digits=5, 
        decimal_places=4,
        description="Expected annual return rate"
    )
    min_investment_amount: Decimal = Field(
        default=Decimal("1.00"),
        max_digits=19,
        decimal_places=4,
        description="Minimum investment amount"
    )
    max_investment_amount: Optional[Decimal] = Field(
        None,
        max_digits=19,
        decimal_places=4,
        description="Maximum investment amount"
    )
    investment_period_days: Optional[int] = Field(
        None, 
        description="Investment period in days (NULL for demand products)"
    )
    description: Optional[str] = Field(None, description="Product description")
    features: Optional[Dict[str, Any]] = Field(None, description="Product features")


class InvestmentProductCreate(InvestmentProductBase):
    """Model for creating investment products."""
    pass


class InvestmentProductUpdate(BaseModel):
    """Model for updating investment products."""
    
    name: Optional[str] = Field(None, max_length=200)
    expected_return_rate: Optional[Decimal] = Field(None, max_digits=5, decimal_places=4)
    min_investment_amount: Optional[Decimal] = Field(None, max_digits=19, decimal_places=4)
    max_investment_amount: Optional[Decimal] = Field(None, max_digits=19, decimal_places=4)
    is_active: Optional[bool] = None
    description: Optional[str] = None
    features: Optional[Dict[str, Any]] = None


class InvestmentProductResponse(InvestmentProductBase):
    """Model for investment product responses."""

    id: UUID = Field(..., description="Product unique identifier")
    is_active: bool = Field(..., description="Whether product is active")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    class Config:
        """Pydantic configuration."""
        from_attributes = True
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


# Risk Assessment models
class RiskAssessmentCreate(BaseModel):
    """Model for creating risk assessments."""
    
    risk_tolerance: RiskLevel = Field(..., description="Risk tolerance level")
    investment_experience: InvestmentExperience = Field(..., description="Investment experience")
    investment_goal: InvestmentGoal = Field(..., description="Investment goal")
    investment_horizon: InvestmentHorizon = Field(..., description="Investment time horizon")
    monthly_income_range: Optional[str] = Field(None, description="Monthly income range")
    assessment_data: Optional[Dict[str, Any]] = Field(None, description="Detailed assessment data")


class RiskAssessmentResponse(RiskAssessmentCreate):
    """Model for risk assessment responses."""
    
    id: UUID = Field(..., description="Assessment unique identifier")
    user_id: UUID = Field(..., description="User ID")
    assessment_score: int = Field(..., description="Calculated assessment score")
    expires_at: datetime = Field(..., description="Assessment expiration date")
    created_at: datetime = Field(..., description="Creation timestamp")
    
    class Config:
        """Pydantic configuration."""
        from_attributes = True


# Investment Transaction models
class InvestmentPurchaseRequest(BaseModel):
    """Model for investment purchase requests."""
    
    account_id: UUID = Field(..., description="Source account ID")
    product_id: UUID = Field(..., description="Investment product ID")
    amount: Decimal = Field(
        ...,
        gt=0,
        max_digits=19,
        decimal_places=4,
        description="Investment amount"
    )
    
    @validator('amount')
    def validate_amount(cls, v):
        """Validate investment amount."""
        if v <= 0:
            raise ValueError('Investment amount must be positive')
        return v


class InvestmentRedemptionRequest(BaseModel):
    """Model for investment redemption requests."""
    
    holding_id: UUID = Field(..., description="Investment holding ID")
    shares: Optional[Decimal] = Field(
        None,
        gt=0,
        max_digits=19,
        decimal_places=8,
        description="Shares to redeem (NULL for full redemption)"
    )
    
    @validator('shares')
    def validate_shares(cls, v):
        """Validate redemption shares."""
        if v is not None and v <= 0:
            raise ValueError('Redemption shares must be positive')
        return v


class InvestmentTransactionResponse(BaseModel):
    """Model for investment transaction responses."""
    
    id: UUID = Field(..., description="Transaction unique identifier")
    user_id: UUID = Field(..., description="User ID")
    account_id: UUID = Field(..., description="Account ID")
    product_id: UUID = Field(..., description="Product ID")
    holding_id: Optional[UUID] = Field(None, description="Holding ID")
    transaction_type: TransactionType = Field(..., description="Transaction type")
    shares: Decimal = Field(..., max_digits=19, decimal_places=8, description="Transaction shares")
    unit_price: Decimal = Field(..., max_digits=19, decimal_places=4, description="Unit price")
    amount: Decimal = Field(..., max_digits=19, decimal_places=4, description="Transaction amount")
    fee: Decimal = Field(..., max_digits=19, decimal_places=4, description="Transaction fee")
    net_amount: Decimal = Field(..., max_digits=19, decimal_places=4, description="Net amount")
    status: TransactionStatus = Field(..., description="Transaction status")
    settlement_date: Optional[datetime] = Field(None, description="Settlement date")
    description: Optional[str] = Field(None, description="Transaction description")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    
    class Config:
        """Pydantic configuration."""
        from_attributes = True
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


# Investment Holding models
class InvestmentHoldingResponse(BaseModel):
    """Model for investment holding responses."""
    
    id: UUID = Field(..., description="Holding unique identifier")
    user_id: UUID = Field(..., description="User ID")
    account_id: UUID = Field(..., description="Account ID")
    product_id: UUID = Field(..., description="Product ID")
    product_name: str = Field(..., description="Product name")
    product_type: ProductType = Field(..., description="Product type")
    shares: Decimal = Field(..., max_digits=19, decimal_places=8, description="Holding shares")
    average_cost: Decimal = Field(..., max_digits=19, decimal_places=4, description="Average cost")
    total_invested: Decimal = Field(..., max_digits=19, decimal_places=4, description="Total invested")
    current_value: Decimal = Field(..., max_digits=19, decimal_places=4, description="Current value")
    unrealized_gain_loss: Decimal = Field(..., max_digits=19, decimal_places=4, description="Unrealized P&L")
    realized_gain_loss: Decimal = Field(..., max_digits=19, decimal_places=4, description="Realized P&L")
    return_rate: Decimal = Field(..., max_digits=8, decimal_places=4, description="Return rate")
    purchase_date: datetime = Field(..., description="Purchase date")
    maturity_date: Optional[datetime] = Field(None, description="Maturity date")
    status: HoldingStatus = Field(..., description="Holding status")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    
    class Config:
        """Pydantic configuration."""
        from_attributes = True
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


# Portfolio Summary models
class PortfolioSummaryResponse(BaseModel):
    """Model for portfolio summary responses."""
    
    total_assets: Decimal = Field(..., max_digits=19, decimal_places=4, description="Total assets")
    total_invested: Decimal = Field(..., max_digits=19, decimal_places=4, description="Total invested")
    total_gain_loss: Decimal = Field(..., max_digits=19, decimal_places=4, description="Total P&L")
    total_return_rate: Decimal = Field(..., max_digits=8, decimal_places=4, description="Total return rate")
    asset_allocation: Dict[ProductType, Decimal] = Field(..., description="Asset allocation by type")
    holdings_count: int = Field(..., description="Number of holdings")
    active_products_count: int = Field(..., description="Number of active products")

    class Config:
        """Pydantic configuration."""
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


# Product NAV models
class ProductNAVResponse(BaseModel):
    """Model for product NAV responses."""
    
    id: UUID = Field(..., description="NAV record unique identifier")
    product_id: UUID = Field(..., description="Product ID")
    nav_date: date = Field(..., description="NAV date")
    unit_nav: Decimal = Field(..., max_digits=19, decimal_places=4, description="Unit NAV")
    accumulated_nav: Optional[Decimal] = Field(None, max_digits=19, decimal_places=4, description="Accumulated NAV")
    daily_return_rate: Optional[Decimal] = Field(None, max_digits=8, decimal_places=6, description="Daily return rate")
    created_at: datetime = Field(..., description="Creation timestamp")
    
    class Config:
        """Pydantic configuration."""
        from_attributes = True
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }


# Product Recommendation models
class ProductRecommendationResponse(BaseModel):
    """Model for product recommendation responses."""
    
    product: InvestmentProductResponse = Field(..., description="Recommended product")
    recommendation_score: float = Field(..., description="Recommendation score (0-1)")
    recommendation_reason: str = Field(..., description="Recommendation reason")
    risk_match: bool = Field(..., description="Whether risk level matches user profile")
    suggested_allocation: Optional[Decimal] = Field(None, description="Suggested allocation percentage")

    class Config:
        """Pydantic configuration."""
        json_encoders = {
            Decimal: lambda v: float(v) if v is not None else None
        }

File: corebank-frontend/src/types/investment.ts
/**
 * Investment-related type definitions for the frontend.
 */

// Enums
export enum ProductType {
  MONEY_FUND = 'money_fund',
  FIXED_TERM = 'fixed_term',
  MUTUAL_FUND = 'mutual_fund',
  INSURANCE = 'insurance'
}

export enum RiskLevel {
  VERY_LOW = 1,
  LOW = 2,
  MEDIUM = 3,
  HIGH = 4,
  VERY_HIGH = 5
}

export enum InvestmentExperience {
  BEGINNER = 'beginner',
  INTERMEDIATE = 'intermediate',
  ADVANCED = 'advanced'
}

export enum InvestmentGoal {
  WEALTH_PRESERVATION = 'wealth_preservation',
  STEADY_GROWTH = 'steady_growth',
  AGGRESSIVE_GROWTH = 'aggressive_growth'
}

export enum InvestmentHorizon {
  SHORT_TERM = 'short_term',
  MEDIUM_TERM = 'medium_term',
  LONG_TERM = 'long_term'
}

export enum TransactionType {
  PURCHASE = 'purchase',
  REDEMPTION = 'redemption',
  DIVIDEND = 'dividend',
  INTEREST = 'interest'
}

export enum TransactionStatus {
  PENDING = 'pending',
  CONFIRMED = 'confirmed',
  FAILED = 'failed',
  CANCELLED = 'cancelled'
}

export enum HoldingStatus {
  ACTIVE = 'active',
  MATURED = 'matured',
  REDEEMED = 'redeemed'
}

// Investment Product interfaces
export interface InvestmentProduct {
  id: string
  product_code: string
  name: string
  product_type: ProductType
  risk_level: RiskLevel
  expected_return_rate?: number
  min_investment_amount: number
  max_investment_amount?: number
  investment_period_days?: number
  is_active: boolean
  description?: string
  features?: Record<string, any>
  created_at: string
  updated_at: string
}

export interface InvestmentProductCreate {
  product_code: string
  name: string
  product_type: ProductType
  risk_level: RiskLevel
  expected_return_rate?: number
  min_investment_amount: number
  max_investment_amount?: number
  investment_period_days?: number
  description?: string
  features?: Record<string, any>
}

// Risk Assessment interfaces
export interface RiskAssessment {
  id: string
  user_id: string
  risk_tolerance: RiskLevel
  investment_experience: InvestmentExperience
  investment_goal: InvestmentGoal
  investment_horizon: InvestmentHorizon
  monthly_income_range?: string
  assessment_score: number
  assessment_data?: Record<string, any>
  expires_at: string
  created_at: string
}

export interface RiskAssessmentCreate {
  risk_tolerance: RiskLevel
  investment_experience: InvestmentExperience
  investment_goal: InvestmentGoal
  investment_horizon: InvestmentHorizon
  monthly_income_range?: string
  assessment_data?: Record<string, any>
}

// Investment Transaction interfaces
export interface InvestmentTransaction {
  id: string
  user_id: string
  account_id: string
  product_id: string
  holding_id?: string
  transaction_type: TransactionType
  shares: number
  unit_price: number
  amount: number
  fee: number
  net_amount: number
  status: TransactionStatus
  settlement_date?: string
  description?: string
  created_at: string
  updated_at: string
}

export interface InvestmentPurchaseRequest {
  account_id: string
  product_id: string
  amount: number
}

export interface InvestmentRedemptionRequest {
  holding_id: string
  shares?: number
}

// Investment Holding interfaces
export interface InvestmentHolding {
  id: string
  user_id: string
  account_id: string
  product_id: string
  product_name: string
  product_type: ProductType
  shares: number
  average_cost: number
  total_invested: number
  current_value: number
  unrealized_gain_loss: number
  realized_gain_loss: number
  return_rate: number
  purchase_date: string
  maturity_date?: string
  status: HoldingStatus
  created_at: string
  updated_at: string
}

// Portfolio interfaces
export interface PortfolioSummary {
  total_assets: number
  total_invested: number
  total_gain_loss: number
  total_return_rate: number
  asset_allocation: Record<ProductType, number>
  holdings_count: number
  active_products_count: number
}

// Product NAV interfaces
export interface ProductNAV {
  id: string
  product_id: string
  nav_date: string
  unit_nav: number
  accumulated_nav?: number
  daily_return_rate?: number
  created_at: string
}

// Product Recommendation interfaces
export interface ProductRecommendation {
  product: InvestmentProduct
  recommendation_score: number
  recommendation_reason: string
  risk_match: boolean
  suggested_allocation?: number
}

// Filter and query interfaces
export interface ProductFilters {
  product_type?: ProductType
  risk_level?: RiskLevel
  is_active?: boolean
  search?: string
}

export interface TransactionFilters {
  product_id?: string
  transaction_type?: TransactionType
  status?: TransactionStatus
  start_date?: string
  end_date?: string
}

// Chart data interfaces
export interface ChartDataPoint {
  date: string
  value: number
  label?: string
}

export interface AssetAllocationData {
  type: ProductType
  value: number
  percentage: number
  color: string
}

export interface PerformanceData {
  period: string
  return_rate: number
  benchmark_return?: number
  volatility?: number
  sharpe_ratio?: number
}

// Form interfaces
export interface RiskAssessmentForm {
  // Personal Information
  age_range: string
  monthly_income_range: string
  investment_experience: InvestmentExperience
  
  // Investment Goals
  investment_goal: InvestmentGoal
  investment_horizon: InvestmentHorizon
  investment_amount_range: string
  
  // Risk Tolerance Questions
  market_decline_reaction: string
  investment_priority: string
  risk_vs_return_preference: string
  liquidity_needs: string
  
  // Additional Information
  other_investments: string[]
  financial_knowledge_level: string
}

// UI State interfaces
export interface InvestmentPageState {
  selectedProduct?: InvestmentProduct
  showPurchaseModal: boolean
  showRedemptionModal: boolean
  showRiskAssessment: boolean
  selectedHolding?: InvestmentHolding
  filters: ProductFilters
  sortBy: string
  sortOrder: 'asc' | 'desc'
}

// API Response interfaces
export interface InvestmentApiResponse<T> {
  data: T
  message?: string
  timestamp: string
}

export interface PaginatedInvestmentResponse<T> {
  items: T[]
  total_count: number
  page: number
  page_size: number
  total_pages: number
}

// Error interfaces
export interface InvestmentError {
  code: string
  message: string
  field?: string
  details?: Record<string, any>
}

// Constants
export const PRODUCT_TYPE_LABELS: Record<ProductType, string> = {
  [ProductType.MONEY_FUND]: '货币基金',
  [ProductType.FIXED_TERM]: '定期理财',
  [ProductType.MUTUAL_FUND]: '基金产品',
  [ProductType.INSURANCE]: '保险理财'
}

export const RISK_LEVEL_LABELS: Record<RiskLevel, string> = {
  [RiskLevel.VERY_LOW]: '极低风险',
  [RiskLevel.LOW]: '低风险',
  [RiskLevel.MEDIUM]: '中等风险',
  [RiskLevel.HIGH]: '高风险',
  [RiskLevel.VERY_HIGH]: '极高风险'
}

export const RISK_LEVEL_COLORS: Record<RiskLevel, string> = {
  [RiskLevel.VERY_LOW]: '#10B981', // green-500
  [RiskLevel.LOW]: '#84CC16',      // lime-500
  [RiskLevel.MEDIUM]: '#F59E0B',   // amber-500
  [RiskLevel.HIGH]: '#EF4444',     // red-500
  [RiskLevel.VERY_HIGH]: '#DC2626' // red-600
}

export const TRANSACTION_TYPE_LABELS: Record<TransactionType, string> = {
  [TransactionType.PURCHASE]: '申购',
  [TransactionType.REDEMPTION]: '赎回',
  [TransactionType.DIVIDEND]: '分红',
  [TransactionType.INTEREST]: '利息'
}

export const TRANSACTION_STATUS_LABELS: Record<TransactionStatus, string> = {
  [TransactionStatus.PENDING]: '处理中',
  [TransactionStatus.CONFIRMED]: '已确认',
  [TransactionStatus.FAILED]: '失败',
  [TransactionStatus.CANCELLED]: '已取消'
}

export const INVESTMENT_EXPERIENCE_LABELS: Record<InvestmentExperience, string> = {
  [InvestmentExperience.BEGINNER]: '初学者',
  [InvestmentExperience.INTERMEDIATE]: '有经验',
  [InvestmentExperience.ADVANCED]: '专业投资者'
}

export const INVESTMENT_GOAL_LABELS: Record<InvestmentGoal, string> = {
  [InvestmentGoal.WEALTH_PRESERVATION]: '财富保值',
  [InvestmentGoal.STEADY_GROWTH]: '稳健增长',
  [InvestmentGoal.AGGRESSIVE_GROWTH]: '积极增长'
}

export const INVESTMENT_HORIZON_LABELS: Record<InvestmentHorizon, string> = {
  [InvestmentHorizon.SHORT_TERM]: '短期（1年以内）',
  [InvestmentHorizon.MEDIUM_TERM]: '中期（1-5年）',
  [InvestmentHorizon.LONG_TERM]: '长期（5年以上）'
}
